extensions/ql-vscode/src/view/webview.tsx

```diff
@@ -25,11 +26,11 @@ const render = () => {
   // eslint-disable-next-line @typescript-eslint/no-var-requires
   const view: WebviewDefinition = require(`./${viewName}/index.tsx`).default;
 
-  ReactDOM_render(
-    view.component,
-    document.getElementById("root"),
-    // Post a message to the extension when fully loaded.
-    () => vscode.postMessage({ t: "viewLoaded", viewName }),
+  createRoot(element).render(
+    // Post a message to the extension when fully loaded. See https://github.com/reactwg/react-18/discussions/5 ("What about the render callback?")
+    <div ref={() => vscode.postMessage({ t: "viewLoaded", viewName })}>
+      {view.component}
+    </div>,
```

Comment:
Shouldn't we be changing `view.component` to accept callbacks and then do:

```
const rootElement = createRoot(element);
rootElement.render(<view.component callback={() => vscode.postMessage({ t: "viewLoaded", viewName })} >)
```

I think at least having the `rootElement` in it's own variable makes it slight easier to parse.

Reply:
I think localising this behaviour to the `webview.tsx` file makes the rest of the components much clearer. I don't think the other components should be concerned with how the `viewLoaded` message gets send and adding an extra `div` shouldn't change anything about the layout.

I will extract the `createRoot` result into a separate variable, this also seems to be shown in [the React documentation](https://beta.reactjs.org/reference/react-dom/client/createRoot#createroot) with a name of `root`. I think `rootElement` could be confusing because you could confuse it with an HTML element. The type name of the return type of `createRoot` is `Root`, so calling it `root` like in the documentation seems like it would make sense.

Reply:
Sounds good, and yeah adding it to view.component doesn't quite make sense. Perhaps we can turn the extra `div` into an `App`-like component, like they have in the docs?